Common recipe

iankronquist opened the pull request:
This branch was rebased off of @tschuy's PR #5 to pick up his changes. Various cherry pickings and force pushes may be necessary once that branch gets merged, but we'll cross that bridge when we come to it.

Right now tests fail and the cookbook doesn't converge, but I'm _so_ close.
The test cookbook doesn't converge because it clones what's fresh, and what's fresh needs python27, epel repos, pgsql and other things to work. This is only an issue when installing dependencies. I would appreciate it if we could figure out how to install dependencies in a more friendly way, or maybe we could clone a different cookbook. Thoughts? I would really like people's input here.
The tests don't pass because serverspec isn't configured correctly. I'll fix that tomorrow.
I think I may rename the resource to default.rb, not common.rb.

@tschuy @jordane @bramwelt I would appreciate your  input and review, as well as anyone else who may chime in.

Commit:
Add test cookbook, runner

Commit:
Update tests to reflect new cookbook

iankronquist committed:
Start defining attributes for common resource

iankronquist committed:
Start defining attributes for common resource

iankronquist committed:
Start defining attributes for common resource

iankronquist committed:
Update common resource according to advice

Thanks @jordane

iankronquist committed:
Update common resource according to advice

Thanks @jordane

iankronquist committed:
Update common resource according to advice

Thanks @jordane

iankronquist committed:
Fix metadata rubocop style errors

iankronquist committed:
Fix Berksfile rubocop style errors

iankronquist committed:
Comment the resources to explain their function

iankronquist committed:
Comment the resources to explain their function

iankronquist committed:
Comment the resources to explain their function

iankronquist committed:
Add requirements_file attribute

If this is specified, the cookbook should try and install this file with pip.
If it is not, the cookbook should install the setup.py.
If the setup.py does not exist, fail.

iankronquist committed:
Add requirements_file attribute

If this is specified, the cookbook should try and install this file with pip.
If it is not, the cookbook should install the setup.py.
If the setup.py does not exist, fail.

iankronquist committed:
Add requirements_file attribute

If this is specified, the cookbook should try and install this file with pip.
If it is not, the cookbook should install the setup.py.
If the setup.py does not exist, fail.

iankronquist committed:
Fix rubocop metadata quote style errors

iankronquist committed:
Fix typo in cookbook metadata

'licens' should be 'license'. Oops, my bad.

iankronquist committed:
Add a provider for the common resource

iankronquist committed:
Add a provider for the common resource

iankronquist committed:
Add a provider for the common resource

iankronquist committed:
Add a mode for the directory git will create

iankronquist committed:
Add a mode for the directory git will create

iankronquist committed:
Add a mode for the directory git will create

iankronquist committed:
Specify behavior when config_template is nil

iankronquist committed:
Specify behavior when config_template is nil

iankronquist committed:
Specify behavior when config_template is nil

iankronquist committed:
Fix common provide rubocop errors

iankronquist committed:
Fix common provide rubocop errors

iankronquist committed:
Remove mode attribute from resource

The git LWRP does not supply us with a mode attribute, so don't try and set one

iankronquist committed:
Remove mode attribute from resource

The git LWRP does not supply us with a mode attribute, so don't try and set one

iankronquist committed:
Foodcritic fixes for common provider

FC009: Resource attribute not recognised: ./providers/common.rb:3
FC017: LWRP does not notify when updated: ./providers/common.rb:1
FC023: Prefer conditional attributes: ./providers/common.rb:13

iankronquist committed:
Foodcritic fixes for common provider

FC009: Resource attribute not recognised: ./providers/common.rb:3
FC017: LWRP does not notify when updated: ./providers/common.rb:1
FC023: Prefer conditional attributes: ./providers/common.rb:13

Commit:
Remove django test recipe

Commit:
Remove useless test files

pop committed:
Rakefile and test push

pop committed:
Added working Rakefile
refs #11
This is essentially the same Rakefile from the chef-cookbooks/yum
repo. There's a bunch of other infrastructure and tests before this
'works' but it's a good first step.

iankronquist committed:
Fix rubocop style errors

* Change syntax for checking for nil
* Change indentation

iankronquist committed:
Fix rubocop style errors

* Change syntax for checking for nil
* Change indentation

iankronquist committed:
Specify git cookbook in metadata

iankronquist committed:
Specify git cookbook in metadata

iankronquist committed:
Don't require git resource in common resource

iankronquist committed:
Don't require git resource in common resource

iankronquist committed:
Change git specific attribute names

git_repo -> repository
git_branch -> revision

iankronquist committed:
Change git specific attribute names

git_repo -> repository
git_branch -> revision

iankronquist committed:
Replace bash resource with execute resource

iankronquist committed:
Replace bash resource with execute resource

iankronquist committed:
Add Busser test runner to dependencies

iankronquist committed:
Add Busser test runner to dependencies

iankronquist committed:
Correct name of resource variables

new_resource.variables -> new_resource.config_vars

iankronquist committed:
Correct name of resource variables

new_resource.variables -> new_resource.config_vars

iankronquist committed:
Remove unnecessary node attribute

iankronquist committed:
Remove unnecessary node attribute

iankronquist committed:
Add git to Berksfile dependencies

iankronquist committed:
Add git to Berksfile dependencies

iankronquist committed:
Create the user if specified

iankronquist committed:
Create the user if specified

iankronquist committed:
Set up directory for checkout with proper perms

iankronquist committed:
Set up directory for checkout with proper perms

iankronquist committed:
Make comparisons in common provider more explicit

iankronquist committed:
Make comparisons in common provider more explicit

iankronquist committed:
Make the default requirements_file an empty string

iankronquist committed:
Make the default requirements_file an empty string

iankronquist committed:
Rename cookbook in metadata

iankronquist committed:
Create a test recipe and accompanying template

iankronquist committed:
Create a test recipe and accompanying template

iankronquist committed:
Fix nil test rubocop error

iankronquist committed:
Fix nil test rubocop error

jordane commented:
@iankronquist you can summon all the chef reviewers with @osuosl-cookbooks/chefs for what it is worth.

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,26 @@
+# If :config_template is nil, do nothing
+attribute :config_template, 'kind_of' => String,
+                            :default => 'settings.py.erb'
+attribute :config_destination, 'kind_of' => String,
```
is this always relative? If so, relative to what?

iankronquist replied:
That's a good point, it shouldn't be relative

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,26 @@
+
+# If this is set, install that requirements file. If it is not, install a
+# setup.py. If the setup.py does not exist, fail.
+attribute :requirements_file, 'kind_of' => String, :default => ''
```
It's much cleaner to do something like:

``` ruby
attribute :requirements_file, 'kind_of' => [String, NilClass], :default => nil
```

Which allows you to just check for nil instead of empty/nonsensical strings.

jordane commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -0,0 +1,48 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include_recipe 'python-webapp'
```
You don't need this to use the `python_webapp_common` resource. That becomes available as soon as you put `depends 'python-webapp'` in the metadata.

In fact, if that line disappears from the metadata, this won't save you.

jordane commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -0,0 +1,48 @@
+# limitations under the License.
+
+include_recipe 'python-webapp'
+include_recipe 'yum'
```
This isn't a `depends` in the metadata, and why is it here anyway?

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,56 @@
+
+  # If a config file template has been specified, create it.
+  template new_resource.config_template do
+    only_if { !new_resource.config_template.nil? }
```
~~Since `:config_template` is forced to be a `String`, it can **never** be `nil`.~~

An update to the `resources/common.rb` file has made this no longer true.

armiller commented on providers/common.rb:
```diff
@@ -0,0 +1,56 @@
+  # Install the application requirements.
+  # If a requirements file has been specified, use pip.
+  # otherwise use the setup.py
+  if new_resource.requirements_file != ''
```
I think `.empty?` would work better than `!= ''`

iankronquist replied:
Even better would probably be @jordane's nil idea.

armiller commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -0,0 +1,48 @@
+  secret_key: 'abcd'
+}
+
+log 'rinning cookbook'
```
rinning cookbook?

pop committed:
inital unit test commit
Included in this commit are the spec_helper.rb and a template spec test
file to base future tests on. The template includes an example as well
as skeleton code.
refs #11

iankronquist committed:
Merge pull request #5 from osuosl-cookbooks/add_testing_infra

Add test cookbook, runner

iankronquist committed:
requirements_file attribute may be nil

This indicates that setup.py should be used.

iankronquist committed:
Config destination should be absolute

iankronquist committed:
config_template may be nil

iankronquist committed:
Remove unnecessary cruft

- Remove debug statement
- Remove unnecessary include_recipe invocations

iankronquist committed:
Merge branch 'common_recipe' of github.com:osuosl-cookbooks/python-webapp-cookbook into common_recipe

Conflicts:
	providers/common.rb
	resources/common.rb
	test/cookbooks/python-webapp-test/metadata.rb
	test/cookbooks/python-webapp-test/recipes/default.rb

iankronquist commented:
So I may have goofed up the merge from master. Let me know if you want to be picky and have me squash or cherry pick to clean up history.
@jordane @armiller thanks for your advice, I addressed your concerns.

ramereth commented on providers/common.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  # Update the code.
+  git new_resource.path do
```
Where do you install git? (the test cookbook should not be where you do it)

ramereth replied:
BTW, its ideal to use the git cookbook.

iankronquist replied:
I depend on the git cookbook in my metadata. I thought that installed git for me, but apparently not.

ramereth replied:
Nope, you still have to use `include_recipe 'git'` somewhere. That just adds the dependency, does not actually execute the recipe.

ramereth commented on providers/common.rb:
```diff
@@ -0,0 +1,56 @@
+  # If a requirements file has been specified, use pip.
+  # otherwise use the setup.py
+  if new_resource.requirements_file.nil?
+    execute 'pip install' do
```
Where do you install pip? I'd assume this can be easily fixed using the python cookbook.

ramereth commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,29 @@
+  config_template 'config.yml.erb'
+  config_destination '/opt/whats_fresh/config.yml'
+  config_vars settings_hash
+  requirements_file nil
```
If the default in the resource is `nil`, why do you need to set it here?

iankronquist replied:
I was thinking that "explicit is better than implicit". I suppose I could delete this.

ramereth replied:
If you really want to test your resource defaults, this would be a good way :smile:

ramereth commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,29 @@
+  group 'whats_fresh'
+
+  repository 'https://github.com/osu-cass/whats-fresh-api.git'
+  revision 'master'
```
If the default is already `master`, why state it here?

ramereth commented on resources/common.rb:
```diff
@@ -0,0 +1,27 @@
+attribute :create_user, 'kind_of' => [TrueClass, FalseClass], :default => false
+
+# Create the owner, path, or group if they do not exist
+attribute :path, 'kind_of' => String, :default => '/opt/app/'
```
Do we really want to define a default path location for applications? Seems like this should be nil by default and be required to be set. Same with the `config_destination`. But maybe I'm wrong.

iankronquist replied:
A default path for applications doesn't seem like a bad thing to have. An application shouldn't care where in the filesystem it lives.
A good example of a path which should be set by default and probably shouldn't be set by the user is the `virtualenv_path`. Nobody cares where that goes.
As for the `config_destination`, it seems useful to have the default be the same as the default for all django apps. You have to go out of your way to move it.

ramereth replied:
But since this is a resource, what happens if you create two resources without changing the default? The better option is to utilize the `name_attribute` as a part of the path (i.e. it should become `/opt/<name>`. Although I'm not sure you do that in the context of an LWRP, but at least you know the default path will always change since we don't have two resources with the same name.

iankronquist replied:
Good point, I'll go fix that.

iankronquist committed:
Add python cookbook dependency

iankronquist committed:
Include git recipe to install git

iankronquist committed:
Use python cookbook for virtualenv and pip lwrps

iankronquist committed:
Remove attributes which were set to default values

iankronquist committed:
Add serverspec integration tests

iankronquist committed:
Use name_attribute to set default app install path

jordane commented on test/integration/default/serverspec/test_git_repo_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  it { should be_directory }
+  it { should be_grouped_into 'chef' }
+  it { should be_readable.by('chef') }
+  it { should be_readable.by('chef') }
```
Duplicate line to L15?

jordane commented on test/integration/default/serverspec/test_template_spec.rb:
```diff
@@ -0,0 +1,8 @@
+describe file('/opt/app/settings.py') do
+  it { should be_grouped_into 'chef' }
+  it { should be_readable.by('chef') }
+  it { should be_readable.by('chef') }
```
Another duplicate line?

jordane commented on test/integration/default/serverspec/test_virtualenv_spec.rb:
```diff
@@ -0,0 +1,9 @@
+  it { should be_directory }
+  it { should be_grouped_into 'chef' }
+  it { should be_readable.by('chef') }
+  it { should be_readable.by('chef') }
```
And another duplicate line?

iankronquist committed:
Remove duplicate lines

iankronquist commented:
Copy paste coding! @jordane fixed.

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,71 @@
+include_recipe 'git'
```
Does this actually work now in Chef 12? Cool.

You should probably use [this](http://www.fewbytes.com/using-include_recipe-in-chef-lwrp/) method for including recipes for Chef 11 compatibility though.

iankronquist replied:
Actually it doesn't work. I was going to ask you about that.

```
       NoMethodError
       -------------
       undefined method `include_recipe' for Chef::Provider::PythonWebappCommon:Class

       Cookbook Trace:
       ---------------
         /tmp/kitchen/cookbooks/python-webapp/providers/common.rb:1:in `class_from_file'

       Relevant File Content:
       ----------------------
       /tmp/kitchen/cookbooks/python-webapp/providers/common.rb:

         1>> include_recipe 'git'
         2:  
         3:  action :install do
         4:  
         5:    if new_resource.create_user
         6:      group new_resource.group do
         7:        action :create
         8:      end
         9:      user new_resource.owner do
        10:        action :create
```

```
```

jordane replied:
Basically, you want `run_context.include_recipe`. The link in my first comment has more details.

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,34 @@
+                            :default => 'settings.py.erb'
+# The destination is an absolute path
+attribute :config_destination, 'kind_of' => String,
+                               :default => '/opt/app/settings.py'
```
This should be changed in regards to the changes made at @ramereth's suggestion about incorporating the resource name in the path.

In effect, this attribute should probably default to `nil`, and then calculate the correct default in the provider.

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,34 @@
+# The destination is an absolute path
+attribute :config_destination, 'kind_of' => String,
+                               :default => '/opt/app/settings.py'
+attribute :config_vars, 'kind_of' => Hash
```
Is this attribute 100% necessary? Otherwise it should have a default.

by 100% necessary I mean does it make sense for a user to not call it.

iankronquist replied:
Maybe you have a template file with no variables?
I'm not sure what a sane default would look like.

jordane replied:
I suspected so, but wanted to confirm. This is fine the way it is then.

jordane commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,27 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+package 'git'
```
This shouldn't be needed anymore?

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  # Create virtual environment
+  python_virtualenv new_resource.virtualenv_path do
```
Does the `python_virtualenv` resource make sure that virtualenv is installed? Otherwise you might want to include the python recipes first.

iankronquist replied:
I think so.

iankronquist replied:
Nope it doesn't, but I fixed it by installing virtualenv.

iankronquist committed:
Changes which should have more specific commit messages

tschuy commented:
Should probably link these on Github. This fixes #6 .

tschuy commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,26 @@
+  create_user true
+  owner 'whats_fresh'
+  group 'whats_fresh'
+
```
This is pretty close to a maximal integration test case already. The only things this does NOT change from the defaults are `path`, `revision`, and `virtualenv_path`. Perhaps consider adding those to this test, and adding another test that tests the default cases?

iankronquist replied:
Actually these are tests of the _defaults_. I agree that we should add another instance for testing changes to these defaults.

iankronquist replied:
I added tests for changing the default `path`, `revision`, and `virtualenv_path`.

jordane commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,26 @@
+
+  config_template 'config.yml.erb'
+  config_destination '/opt/whats_fresh/config.yml'
+  config_vars settings_hash
```
If you want to construct the hash here, do

``` ruby
config_vars(
  host: 'db.example.com',
  port: 5432,
  ...
)
```

tschuy commented:
This resource currently doesn't handle python package installation as an attribute.

iankronquist commented:
@tschuy I don't understand what you mean. What would you like to see?

iankronquist committed:
Change file header

iankronquist committed:
Fix typo, destination should be default

iankronquist committed:
Include pip and virtualenv recipes

This so pip and virtualenv are installed

iankronquist committed:
Fix logic error

I changed a falsey-ness check into an explicit nil? check without changing
logic. Oops!

iankronquist committed:
Don't install git in the recipe

iankronquist committed:
Move hash within LWRP

iankronquist committed:
Change quoting style to satisfy rubocop

iankronquist committed:
Update tests to reflect changes in reality

iankronquist committed:
Include path in configuration file's destination

iankronquist commented:
Okay, I think this is ready. All tests pass. I'm not sure what @tschuy thinks should change.
I will not be in tomorrow, or at the summit, sorry. I'll also be in Portland for Spring Break. Feel free to update this PR as necessary to get it merged. Of course I'll be willing to do some work over the break, ping me on IRC or by comments here.

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,83 @@
+
+  # If the new_resource.path is nil set the install path to
+  # `/opt/name_attribute`
+  if new_resource.path.nil?
```
A more idiomatic ruby way of doing this is using `or`

``` ruby
path = new_resource.path or "/opt/#{new_resource.name}"
```

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,83 @@
+
+  # If the config_destination is nil, as by default, default to
+  # /opt/<path>/settings.py
+  if new_resource.config_destination.nil?
```
Same thing here, you can just use `or`

iankronquist replied:
So if I use or it goes over 80 characters and rubocop whines at me and recommends using an if statement. Is there a better way of doing it which doesn't go over 80 characters?

jordane replied:
Ah, I thought this would fit in under 80 characters.

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,83 @@
+    config_destination = new_resource.config_destination
+  end
+
+  directory path do
```
You should also set `recursive true` for this resource.

iankronquist replied:
Done.

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,34 @@
+default_action :install
+
+# Name the resource
+attribute :name, :name_attribute => true, 'kind_of' => String,
```
There is no need to redefine the name attribute, all resources get this one for free.

jordane commented on resources/common.rb:
```diff
@@ -0,0 +1,34 @@
+attribute :revision, 'kind_of' => String, :default => 'master'
+
+# Location of the virtualenv
+attribute :virtualenv_path, 'kind_of' => String, :default => '/opt/venv'
```
`/opt/venv` seems like a terrible default to me.

iankronquist replied:
How about `/opt/venv_$PROJECT_NAME` and have this nil or String nonsense further infect the project?

jordane replied:
Yep, that works.

jordane commented on providers/common.rb:
```diff
@@ -0,0 +1,83 @@
+  # Install the application requirements.
+  # If a requirements file has been specified, use pip.
+  # otherwise use the setup.py
+  if !new_resource.requirements_file.nil?
```
Usually `unless` is used rather than `if !`

iankronquist replied:
Rubocop says that I shouldn't use unless with an else block, I should just put the positive case first, which is what I'm doing.

jordane replied:
Ah, that's an odd rule, but okay.

iankronquist committed:
Make common provider more like idiomatic ruby

iankronquist committed:
Remove name attribute - it comes free with chef

iankronquist committed:
Fix typo

nil should be NilClass.

iankronquist committed:
Change destination of the config file

iankronquist committed:
Add an invocation of the LWRP to test non-defaults

iankronquist committed:
Change location of config file in test

iankronquist committed:
Test new LWRP invocation

iankronquist committed:
Add test config template

tschuy commented:
@iankronquist in the Django issue, you say that installing additional pip packages shouldn't be handled by the Django provider. Should it be provided here as an attribute, or do we never want people to be able to install add'l pip packages?

tschuy commented:
@iankronquist looking at the list of "settings you're setting" vs "settings with defaults", you're setting name, create_user, owner, group, config_template, config_destination, and config_vars to non-default values. we should have something with those set as defaults, shouldn't we?

iankronquist commented:
@tschuy You mean pip packages in addition to what's in the `requirements.txt`? What's the use case? It isn't Django specific, so if we need it, it belongs in here, but I'm not convinced it's necessary.

I should probably update the tests to be more comprehensive.

iankronquist committed:
Make virtualenv path default to venv_$PROJECT_NAME

iankronquist committed:
Test default and non-default settings

iankronquist commented:
Alright, @tschuy @jordane I'm going to be sporadically working on this PR this week. It's ready for (more) review.

iankronquist committed:
Require the yum cookbook in the test cookbook

iankronquist committed:
Install system requirements needed for whats_fresh

iankronquist commented:
@jordane @tschuy Can we get this reviewed?

Commit:
Update readme and gemfile for running tests

Commit:
Set up more sane Rakefile configuration

Commit:
Explain test running in README, make Travis use rake

iankronquist committed:
Move common.rb files to default.rb

iankronquist committed:
Add epel and ius cookbook dependencies

iankronquist committed:
Use python-test-cookbooks instead of WW and WF

Commit:
Remove rake tags

Commit:
Update rakefile to pass its own tests

iankronquist committed:
Update tests for python-test-cookbooks

iankronquist committed:
Test new attributes

iankronquist committed:
Fix FoodCritic errors

iankronquist committed:
Fix rubocop errors

Commit:
Merge branch 'elijah-testing' into common_recipe

Conflicts:
	Gemfile

Commit:
Merge branch 'common_recipe' of github.com:osuosl-cookbooks/python-webapp into common_recipe

Commit:
Set Travis to use Ruby 2.1

Commit:
Remove blank line

tschuy commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,51 @@
+}
+
+python_webapp 'whats_fresh' do
+  create_user true
```
why do you need to manually create the above user and group?

tschuy replied:
Ah, because the default is "chef"/"chef", and python_webapp doesn't create by default.

We might want to have a special case where, if user/group aren't set, and create_user isn't set, we create the chef/chef anyway?

iankronquist replied:
I prefer the 'treat the programmer as an adult' method and do not do anything unless explicitly instructed.

Commit:
Begin writing unit tests

Commit:
Continue work on unit testing

Commit:
Remove extraneous slash

iankronquist committed:
Fix rubocop error -- line too far indented

Commit:
Add third usage of python_webapp for unit testing

Commit:
Merge branch 'common_recipe' of github.com:osuosl-cookbooks/python-webapp into common_recipe

Commit:
Remove extraneous line

iankronquist committed:
Test the interpreter is chosen properly

iankronquist committed:
Expect that the config template is not created

iankronquist committed:
Expect that requirements.txt is installed with pip

Commit:
Update tests with extra test case

iankronquist committed:
Fix merge conflicts

Commit:
Install What's Fresh with setup.py

iankronquist committed:
Fix long lines

iankronquist committed:
Merge branch 'common_recipe' of github.com:osuosl-cookbooks/python-webapp-cookbook into common_recipe

Commit:
Fix tests and rubocop errors

Commit:
Test Git revision

tschuy commented:
@jordane @osuosl-cookbooks/chefs We've added test coverage for the LWRP, and are pretty sure it's good to go. Can you all take another look at it?

iankronquist committed:
Fix indentation to make rubocop happy

bramwelt commented on Berksfile:
```diff
@@ -2,3 +2,5 @@ source 'https://supermarket.getchef.com'
 
 cookbook 'python-webapp-test', path: 'test/cookbooks/python-webapp-test'
 metadata
+cookbook 'git'
```
Unless these are coming from somewhere other than supermarket, they don't need to be listed here as they are already listed in metadata.

bramwelt commented on README.md:
```diff
@@ -9,36 +9,50 @@ Design documentation is currently in a
+integration, style, and unit testing from one tool.
 
-## Usage
+To run all tests, if you've set up your cloud environment, run:
```
Setup my cloud environment?

bramwelt commented on README.md:
````diff
@@ -9,36 +9,50 @@ Design documentation is currently in a
+Otherwise, you can run all tests using a Vagrant virtual machine with:
+
+```
+$ rake
````
Perhaps this should come first, followed by "If you have access to an openstack instance (etc.) you can use `rake cloud`"

bramwelt commented on Rakefile:
```diff
@@ -0,0 +1,54 @@
+task cloud: ['style', 'spec', 'integration:cloud']
+
+desc 'Run style and spec tests (for Travis)'
+task travis: %w(style spec)
```
This seems inconsistent style-wise with the other task definitions. They should all be switched to using the `%w( )` format or this should be switched to a list.

bramwelt commented on providers/default.rb:
```diff
@@ -0,0 +1,113 @@
+action :install do
```
Missing License Header

tschuy replied:
I've added the header.

bramwelt commented on resources/default.rb:
```diff
@@ -0,0 +1,38 @@
+default_action :install
```
Missing License Header

tschuy replied:
I've added the header.

Commit:
Remove unnecessary dependencies

Commit:
Rewrite test running instructions in README

Commit:
Equalize list style in Rakefile

Commit:
Add proper license headers

Commit:
Add Apache header to Rakefil

tschuy commented on Rakefile:
```diff
@@ -0,0 +1,74 @@
+# Cookbook Name:: osl_application_python_test
+# Recipe:: default
+#
+# Copyright Chef, OSU Open Source Lab
```
Is this the right way to use a single file from someone else's repository?

iankronquist replied:
Nope.
From the Apache 2 license:

```
4. Redistribution. You may reproduce and distribute copies of the
Work or Derivative Works thereof in any medium, with or without
modifications, and in Source or Object form, provided that You
meet the following conditions:

(a) You must give any other recipients of the Work or
Derivative Works a copy of this License; and

(b) You must cause any modified files to carry prominent notices
stating that You changed the files; and

(c) You must retain, in the Source form of any Derivative Works
that You distribute, all copyright, patent, trademark, and
attribution notices from the Source form of the Work,
excluding those notices that do not pertain to any part of
the Derivative Works; and

(d) If the Work includes a "NOTICE" text file as part of its
distribution, then any Derivative Works that You distribute must
include a readable copy of the attribution notices contained
within such NOTICE file, excluding those notices that do not
pertain to any part of the Derivative Works, in at least one
of the following places: within a NOTICE text file distributed
as part of the Derivative Works; within the Source form or
documentation, if provided along with the Derivative Works; or,
within a display generated by the Derivative Works, if and
wherever such third-party notices normally appear. The contents
of the NOTICE file are for informational purposes only and
do not modify the License. You may add Your own attribution
notices within Derivative Works that You distribute, alongside
or as an addendum to the NOTICE text from the Work, provided
that such additional attribution notices cannot be construed
as modifying the License.
You may add Your own copyright statement to Your modifications and
may provide additional or different license terms and conditions
for use, reproduction, or distribution of Your modifications, or
for any such Derivative Works as a whole, provided Your use,
reproduction, and distribution of the Work otherwise complies with
the conditions stated in this License.
```

iankronquist committed:
Merge branch 'common_recipe' of github.com:osuosl-cookbooks/python-webapp-cookbook into common_recipe

tschuy commented:
Ah. I'll fix that up in a bit.

Commit:
Add failing tests for supervisor and gunicorn

Commit:
Add failing tests for supervisor and gunicorn

Commit:
Add failing tests for supervisor and gunicorn

Commit:
Add failing tests for supervisor and gunicorn

Commit:
Add integration tests for gunicorn listening, supervisor service

Commit:
Add integration tests for gunicorn listening, supervisor service

Commit:
Add integration tests for gunicorn listening, supervisor service

Commit:
Flesh out gunicorn unit tests

Commit:
Flesh out gunicorn unit tests

Commit:
Flesh out gunicorn unit tests

Commit:
Add gunicorn and supervisor

Commit:
Add gunicorn and supervisor

Commit:
Add gunicorn and supervisor

tschuy commented:
@iankronquist what's wrong with how I was referencing the original? I linked their source code, put them in the copyright message, ...

iankronquist commented:
@tschuy actually it looks like I just read the part I could see in the comments. What you have looks fine, except that the cookbook name is wrong, and it is not part of a recipe.

iankronquist committed:
Fix Rakefile header chef metadata

iankronquist commented:
@tschuy fixed.

iankronquist committed:
Remove common resource and provider

I think these snuck in by accident.

iankronquist committed:
Refactor whats_fresh and working_waterfronts

iankronquist committed:
Refactor out whats_fresh and working_waterfronts

tschuy commented on test/cookbooks/python-webapp-test/recipes/default.rb:
```diff
@@ -15,3 +15,51 @@
+end
+
+python_webapp 'tutorial-b' do
+  path '/opt/tutorial-b'
```
You should set this to a path that isn't just /opt/(name of resource). that's why we have this one :)

Commit:
Remove references to real projects

Commit:
Remove references to real projects

Commit:
Make tutorial names lowercase

Commit:
Make tutorial names lowercase

Commit:
Fix git merge mistake

Commit:
Make the Vagrant task runnable in the Rakefile

Commit:
Rename expected values from /opt to /app

We should be testing a non-default location for the app.

Commit:
Rename /opt/tutorial-b to /app/tutorial-b

We should be testing a non-default location for the app.

iankronquist committed:
Merge branch 'gunicorn' of github.com:osuosl-cookbooks/python-webapp-cookbook into gunicorn

Conflicts:
	spec/unit/default_spec.rb
	test/integration/default/serverspec/test_supervisor_spec.rb

tschuy commented on Rakefile:
```diff
@@ -0,0 +1,76 @@
+task cloud: %w(style spec integration:cloud)
+
+desc 'Run all tests (style, spec, integration) on Vagrant'
+task vagrant: %w(style spec integration:vagrant)
```
This is the default task, as documented in the README. I suppose it could've had a better 'desc', or one at all.

iankronquist replied:
When I type `rake vagrant` I expect this test to be run.

tschuy replied:
Ok. Better to be explicit than implicit, we want the ability to change the default without breaking things :)

iankronquist committed:
Merge branch 'gunicorn' of github.com:osuosl-cookbooks/python-webapp-cookbook into gunicorn

Conflicts:
	spec/unit/default_spec.rb

iankronquist committed:
Remove empty line

ramereth commented on Rakefile:
```diff
@@ -0,0 +1,76 @@
+#
+# Cookbook Name:: python-webapp
+#
+# Copyright Chef, OSU Open Source Lab
```
This should be changed from `OSU Open Source Lab` to `Oregon State University`. Please change this in all places referenced in the copyright header.

tschuy committed:
Merge pull request #20 from osuosl-cookbooks/gunicorn

Add Gunicorn to LWRP

Commit:
Change copyright message to Oregon State Uni.

tschuy commented:
We should set SELinux if it exists so that nginx can connect to gunicorn:

`node['selinux']['booleans']['httpd_can_network_connect'] = true`

jordane commented on .travis.yml:
```diff
@@ -1,8 +1,9 @@
 script:
-  - bundle exec rubocop .
-  - bundle exec foodcritic .
+  - rake travis
```
It is good practice to actually do `bundle exec rake travis` here.

jordane commented on providers/default.rb:
```diff
@@ -0,0 +1,158 @@
+    config_destination = new_resource.config_destination
+  end
+
+  directory path do
```
The default of `nil` will make this explode. You may want to do a check above and then a `Chef::Log` to inform the user of their mistake.

iankronquist replied:
It passes tests without exploding? I'm not sure what sort of a mistake you're referring to.

iankronquist replied:
This has already been covered.

jordane commented on providers/default.rb:
```diff
@@ -0,0 +1,158 @@
+  end
+
+  # Run django migrations if the django_migrate flag is set
+  bash "run migrations #{new_resource.name}" do
```
Are migrations idempotent? One issue is this will always get run, on every chef run, if `django_migrate` is set to true.

iankronquist replied:
The migrate command with south is idempotent.
https://groups.google.com/forum/#!topic/south-users/9TJuEf-gJEE
That's not an answer.
@tschuy ?

iankronquist replied:
From the `#django` channel:

```
19:19:03        muricula | Are migrations idempotent?
19:20:05      barraponto | are they?                                                                 
19:22:58        FunkyBob | muricula: there is a log kept in the DB to record which wre applied
19:23:30    +SmileyChris | muricula: so the migration scripts themselves aren't, but the migration
                         | function is  
```

So, yes what we're doing is idempotent.

jordane commented on providers/default.rb:
```diff
@@ -0,0 +1,158 @@
+  end
+
+  if new_resource.gunicorn_port
+    # required to work around https://github.com/poise/python/issues/100
```
You use `python_pip` above. Don't you also need this work around there?

iankronquist replied:
How would you use pip to run the `manage.py` script? That doesn't make sense to me, and I don't see a way with the pip provider.

jordane replied:
Further above. You are already calling `python_pip`, but you don't have this workaround if it is needed.

jordane replied:
As in [here](https://github.com/osuosl-cookbooks/python-webapp/pull/14/files#diff-a27dfd053e213b0e3bff4c49b8a9894eR103)

jordane commented on providers/default.rb:
```diff
@@ -0,0 +1,158 @@
+
+  if new_resource.gunicorn_port
+    # required to work around https://github.com/poise/python/issues/100
+    bash 'manually upgrade setuptools' do
```
You can do this with the `python_pip` resource. That would bne better.

iankronquist committed:
Use bundle exec in travis

iankronquist committed:
Use python_pip resource to upgrade setuptools

Replace the bash resource with python_pip resource.

tschuy commented on providers/default.rb:
```diff
@@ -0,0 +1,157 @@
+    only_if { !new_resource.config_template.nil? }
+    action :create
+    source new_resource.config_template
+    path config_destination
```
We should create this path recursively. I want to be able to put my config in /opt/whats_fresh/config/config.yml without having to create /opt/whats_fresh/config/ first.

tschuy replied:
Actually, it appears doing this would be non-trivial (as you'd have to pass in the directory and filenames separately, or separate them with regexfoo). We should leave this up to the user.

iankronquist committed:
Remove Centos 6 specific code

iankronquist committed:
Add a Centos 6 specific recipe

iankronquist committed:
Add CentOs 7 and Ubuntu to test kitchen

Also add a Centos 6.6 specific recipe to that platform runlist.

jordane commented on README.md:
```diff
@@ -7,42 +7,62 @@ Design documentation is currently in a
 ## Supported Platforms
 
 Centos 6
+
```
Why is this newline here?

tschuy commented:
The more I play around with this, the more annoying putting the git clone in /opt/<app name> gets. 

The git clone should go in /opt/<app name>/source, and the venv in /opt/<app name>/venv. This frees us up to do things like put the gunicorn_config.py _somewhere that isn't in a git repo, which is a grade A bad idea_, and allows us to put config in /opt/<app name>/config/ if we want.

mathuin commented:
Is a virtualenv a requirement for compliance with this cookbook?

tschuy commented:
@mathuin as discussed in IRC, not requiring virtualenv shouldn't be too difficult to accomplish but is not a criteria for the initial release of this cookbook. Issue #25 is open and I'd be happy to look at a PR.

mathuin commented:
Okay, I will make a PR based on `common_recipe`.  If this PR gets merged before I'm done, I'll rebase off `master`.

iankronquist committed:
Add ubuntu 12.04 platform to test kitchen

Commit:
Update unit tests for directory shuffle

Commit:
Update integration tests for directory shuffle

Commit:
Do the directory shuffle

iankronquist committed:
Add apt cookbook to Berksfile & Ubuntu 12 runlist

iankronquist committed:
Fix typo in .kitchen.cloud.yml

tschuy commented:
I'd like #26 to get merged into this branch before this branch gets merged.

iankronquist committed:
Include new platforms in README

tschuy committed:
Merge pull request #26 from osuosl-cookbooks/directory_shuffle

Do the directory shuffle

iankronquist commented:
@jordane Can you take a look at this? I'd like to get it merged.

jordane commented:
+1
